lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -432,6 +424,82 @@ _estimateMaxNumberOfInputs txMaxSize ctx outs =
         size = estimateTxSize (txWitnessTagFor @k) ctx sel
         sel  = dummySkeleton (fromIntegral nInps) outs
 
+_initSelectionCriteria
+    :: forall k. TxWitnessTagFor k
+    => ProtocolParameters
+    -> TransactionCtx
+    -> UTxOIndex.UTxOIndex
+    -> NE.NonEmpty TxOut
+    -> Either ErrSelectionCriteria SelectionCriteria
+_initSelectionCriteria pp ctx utxoAvailable outputsUnprepared
+    | (address, assetCount) : _ <- excessivelyLargeBundles =
+        Left $
+            -- We encountered one or more excessively large token bundles.
+            -- Just report the first such bundle:
+            ErrSelectionCriteriaOutputTokenBundleSizeExceedsLimit $
+            ErrOutputTokenBundleSizeExceedsLimit {address, assetCount}
+    | (address, asset, quantity) : _ <- excessiveTokenQuantities =
+        Left $
+            -- We encountered one or more excessive token quantities.
+            -- Just report the first such quantity:
+            ErrSelectionCriteriaOutputTokenQuantityExceedsLimit $
```

Comment:
Just wondering: did you consider adding this logic to `performSelection` instead?

Reply:
> Just wondering: did you consider adding this logic to `performSelection` instead?

That's a good point. The `performSelection` function would also be a perfectly valid place to put this validation logic.

However, doing so would mean a little more work, as we'd have to adjust the property tests for `performSelection`. It might not be too hard though.